### PR TITLE
HPCC-19006 Initial support for a cache for parsed ECL definitions

### PR DESCRIPTION
**Pull request by ghalliday**
If the cache is up to date, then generating an archive will avoid
parsing any of the ECL code.

Also adds a new eclcc option --expand to expand an archive to a directory.

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
HPCC-19005 added to document new eclcc option.
@shamser please review

**Comment by ghalliday** [comment]
my branch issue17472b contains the same code, but with the original commits in place if you want to follow the flow of implementation.

**Comment by ghalliday** [comment]
NOTE: There is a linked change in the LN repo: https://github.com/hpcc-systems/LN/pull/871

**Review by shamser — COMMENTED** [review]
Reviewed.  Just a few minor comments.

File: ecl/hql/hqlexpr.cpp [hunk 1]
```diff
@@ -1060,49 +1069,187 @@ bool HqlParseContext::checkEndMeta()
...
+    //If after finishing this item we will be back to the global stack, then start gathering again
+    metaState.gatherNow = (metaStack.ordinality() == 2);
     return wasGathering;
 }
 
-void HqlParseContext::finishMeta(bool isSeparateFile, bool success)
+void HqlParseContext::finishMeta(bool isSeparateFile, bool success, bool generateMeta, bool canCache, bool isMacro, IHqlExpression * simplifiedDefinition)
 {
-    if (metaState.nesting.empty())  // paranoid - could only happen on an internal error
+    if (metaStack.empty())  // paranoid - could only happen on an internal error
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply ecl/hql/hqlexpr.cpp]: no - something else will have already have been reported.  The intent is to protect the following code from exceptional situations.

File: ecl/hql/hqlexpr.cpp [hunk 2]
```diff
@@ -1060,49 +1069,187 @@ bool HqlParseContext::checkEndMeta()
...
+        if (fullName && !hasPrefix(fullName, INTERNAL_LOCAL_MODULE_NAME, true))
+        {
+            baseFilename.append(metaOptions.cacheLocation);
+            addPathSepChar(baseFilename);
+            convertSelectsToPath(baseFilename, fullName);
+            recursiveCreateDirectoryForFile(baseFilename);
+
+            Owned<IEclCachedDefinition> cached = cache->getDefinition(fullName);
+            if (cached->isUpToDate(optionHash))
+                createCacheEntry = false;
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply ecl/hql/hqlexpr.cpp]: It would need to be on an option since there may be several thousand items, but sounds like a good idea.

**Comment by ghalliday** [comment]
Hopefully the latest version will remove the build problems.

**Review by shamser — APPROVED** [review]
@richardkchapman Code reviewed. Looks good.

**Comment by richardkchapman** [comment]
@ghalliday Do you want to squash this before I merge it?

**Comment by ghalliday** [comment]
@richardkchapman squashed.  Note there is a LN PR as well.

**Pull request merged**